src/tools/dataset.cpp

```diff
@@ -1253,4 +1348,394 @@ void ustore_docs_export(ustore_docs_export_t* c_ptr) noexcept(false) {
+                edges[csv_idx] = data[idx + 2];
+            }
+            else {
+                auto str = fmt::format( //
+                    "{{\"{}\":{},\"{}\":{},\"{}\":{}}}\n",
+                    c.source_id_field,
+                    data[idx],
+                    c.target_id_field,
+                    data[idx + 1],
+                    c.edge_id_field,
+                    data[idx + 2]);
+                write(handle, str.data(), str.size());
+            }
+        }
+    }
+    status = sources_builder.AppendValues(sources.begin(), csv_idx);
+    return_error_if_m(status.ok(), c.error, 0, "Can't append sources");
+    status = targets_builder.AppendValues(targets.begin(), csv_idx);
+    return_error_if_m(status.ok(), c.error, 0, "Can't append targets");
+    if (c.edge_id_field) {
+        status = edges_builder.AppendValues(edges.begin(), csv_idx);
+        return_error_if_m(status.ok(), c.error, 0, "Can't append edges");
+    }
+}
+
+void end_csv_graph( //
+    ustore_graph_export_t& c,
+    int_builder_t& sources_builder,
+    int_builder_t& targets_builder,
+    int_builder_t& edges_builder) {
+
+    array_t sources_array;
+    array_t targets_array;
+    array_t edges_array;
+
+    auto status = sources_builder.Finish(&sources_array);
+    return_error_if_m(status.ok(), c.error, 0, "Can't finish array(sources)");
+    status = targets_builder.Finish(&targets_array);
+    return_error_if_m(status.ok(), c.error, 0, "Can't finish array(targets)");
+    if (c.edge_id_field) {
+        status = edges_builder.Finish(&edges_array);
+        return_error_if_m(status.ok(), c.error, 0, "Can't finish array(edges)");
+    }
+
+    arrow::FieldVector fields;
+    fields.push_back(arrow::field(c.source_id_field, arrow::int64()));
+    fields.push_back(arrow::field(c.target_id_field, arrow::int64()));
+    if (c.edge_id_field)
+        fields.push_back(arrow::field(c.edge_id_field, arrow::int64()));
+
+    std::shared_ptr<arrow::Schema> schema = std::make_shared<arrow::Schema>(fields);
+    std::shared_ptr<arrow::Table> table = nullptr;
+
+    if (c.edge_id_field)
+        table = arrow::Table::Make(schema, {sources_array, targets_array, edges_array});
+    else
+        table = arrow::Table::Make(schema, {sources_array, targets_array});
+    return_error_if_m(table, c.error, 0, "Can't make schema");
+
+    auto maybe_outstream =
+        arrow::io::FileOutputStream::Open(fmt::format("{}{}", generate_file_name(), c.paths_extension));
+    return_error_if_m(maybe_outstream.ok(), c.error, 0, "Can't open file");
+    std::shared_ptr<arrow::io::FileOutputStream> outstream = *maybe_outstream;
+
+    status = arrow::csv::WriteCSV(*table, arrow::csv::WriteOptions::Defaults(), outstream.get());
+    return_error_if_m(status.ok(), c.error, 0, "Can't write in file");
+}
+
+void end_ndjson(int fd) {
+    close(fd);
+}
+
+#pragma region - Main Functions(Graph)
+
+void ustore_graph_import(ustore_graph_import_t* c_ptr) noexcept(false) {
```

Comment:
This is a linking issue! C99 function can't be `noexcept(false)`.

Reply:
We use the functionality of c++11 and above.

Reply:
Fixed.